perf: faster legacy table chart

ktmud opened the pull request:
This commit tries to optimize the performance of the legacy data table.

1. Still use the jquery.databables plugin.
2. Instead of `d3` and `reactify`, we create the DOM with React itself and initialize `jquery.DataTables` with a `useEffect` hook.
3. Removed dependency on `dompurify`. Now that HTML tags will be rendered as plain text (by React) instead of `innerHTML`.
4. Removed a couple of not-so-popular features. See notes below.
5. Added a script to build only specific plugin for faster development.
6. Restyled the pagination so it looks nicer in dashboards.

Plus some minor changes to fix linting and building.

Related: https://github.com/apache/incubator-superset/pull/9234

## Screenshots

Tested with a ~10,000 rows, 7 columns table in Superset, pagination enabled:

### Before 

![legacy-table--old](https://user-images.githubusercontent.com/335541/75824956-4b98d080-5d59-11ea-9f86-4a17d0770e6c.gif)

Takes about 6 secs to load.

![Snip20200303_3](https://user-images.githubusercontent.com/335541/75825254-d8dc2500-5d59-11ea-8da1-8091c8afcc8b.png)


### After

![legacy-table--new](https://user-images.githubusercontent.com/335541/75824966-4dfb2a80-5d59-11ea-83a4-0a3468f9f7c4.gif)

Takes only 3 secs to load!

![Snip20200303_4](https://user-images.githubusercontent.com/335541/75825271-dda0d900-5d59-11ea-98a9-e04fc2445aac.png)



🏆 Enhancements

ktmud committed:
feat: faster legacy table chart

This commit tries to optimize the performance of the legacy data table.

1. Converting everything to Typescript.
2. Create a native React component instead of `reactify` (although all
   DOM operaions still happen in the jQuery.DataTables plugin.
3. Remove dependency on d3, optimize how bars are rendered.

Plus some minor changes to fix linting and building.

Also added a script to build only specific plugin for faster
development.

ktmud commented on package.json:
```diff
@@ -159,5 +159,8 @@
     ]
+  },
+  "resolutions": {
+    "**/@types/react": "^16.9.23"
```
Had to specify `@types/react` version because `interweave-emoji` from `lunar` somehow [picked a fixed version](https://github.com/milesj/interweave/commit/0418313060f88b196ed59dfc0c5b55b6028205a5#r37505021) and it's throwing errors in the latest TypeScript.

ktmud commented on packages/superset-ui-plugins-demo/package.json:
```diff
@@ -39,7 +39,8 @@
     "@types/react": "^16.8.8",
     "@types/storybook__react": "4.0.2",
-    "bootstrap": "^4.3.1",
+    "bootstrap": "^3.4.1",
```
Current Superset is still on Bootstrap 3.

ktmud commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,231 @@
+    // TODO: add back the broken dashboard filters feature
+    // filters = {},
+    // onAddFilter = NOOP,
+    // onRemoveFilter = NOOP,
```
The table filters feature (click on table cells to add a filter) is still broken in production. Discussed with @graceguo-supercat , it's better to skip implementing this feature for now. Might add back later in another PR.

ktmud commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,231 @@
+    // onAddFilter = NOOP,
+    // onRemoveFilter = NOOP,
+    // tableFilter,
+    // timeseriesLimitMetric,
```
`timeseriesLimitMetric` is used to sort table by timestamp when present. Didn't think this is very useful (there are many ways to sort the input data already), so didn't implement in this PR either.

ktmud commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,231 @@
+  });
+
+  const viewportHeight = Math.min(height, window.innerHeight);
+  const pageLengthChoices = [10, 25, 40, 50, 75, 100, 150, 200];
```
This is in consistent with [superset-frontend/src/explore/controls.js](https://github.com/apache/incubator-superset/blob/291306392443a5a0d0e2ee0cc4a95d37c56d4589/superset-frontend/src/explore/controls.jsx#L1336). Hard-code for now, constantize when refactoring controls to plugins.

ktmud commented on packages/superset-ui-plugins-demo/.storybook/storybook.css:
```diff
@@ -6,3 +6,7 @@ body,
+
+#root > div {
+  padding: 8px;
+}
```
Add a padding so it looks more close to an embedded Superset chart in dashboards.

ktmud commented on packages/superset-ui-plugins-demo/storybook/stories/legacy-plugin-chart-table/Stories.tsx:
```diff
@@ -1,9 +1,22 @@
 import { SuperChart } from '@superset-ui/chart';
+import { Props as SuperChartProps } from '@superset-ui/chart/lib/components/SuperChart';
 import data from './data';
+import birthNames from './birth_names.json';
```
Large example data queried from Superset example dataset populated by `superset load_examples`.

ktmud commented on packages/superset-ui-preset-chart-xy/src/encodeable/AxisAgent.ts:
```diff
@@ -133,7 +133,7 @@ export default class AxisAgent<Def extends ChannelDef> {
     const { labelOverlap, labelPadding, orient } = this.config;
 
-    const maxWidth = Math.max(...tickLabelDimensions.map(d => d.width), 0);
+    const maxWidth = Math.max(...tickLabelDimensions.map((d: Dimension) => d.width), 0);
```
Needed to pass Typescript linting.

ktmud commented on packages/superset-ui-legacy-plugin-chart-table/src/Table.css:
```diff
@@ -17,17 +17,25 @@
  */
 .superset-legacy-chart-table {
-  margin: 0px !important;
+  margin: 0px auto;
```
Not sure why there is an `!important` here but it doesn't seem warranted. Tested in Superset and it worked fine.

ktmud reviewed (commented):
Some notes.

ktmud committed:
feat(legacy-table-chart): use React to render DOM

Unfortunately jquery.datatables uses innerHTML to create cell content,
and all rows are created even with pagination.

https://github.com/DataTables/DataTables/blob/83657a29e33ce93ee940ce25684940eb3acb2913/media/js/jquery.dataTables.js#L3113-L3117

This is slow and insecure. We are reverting to DOM data source as in
previous implementation, but instead of using D3 to create the table rows,
we use React. This also renders `dompurify.sanitize` unnecessary since
React will take care of it.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -177,22 +151,28 @@ export default function ReactDataTable(props: DataTableProps) {
+      return formatTimestamp(val);
+    } else if (typeof val === 'string') {
+      // It's fine to return raw value react will handles HTML escaping
+      return val;
```
Could we also test if the cells can render html links correctly?

ktmud replied:
Updated with [js-xss](https://github.com/leizongmin/js-xss) library.

kristw reviewed (commented):
> Removed dependency on `dompurify`. Now that HTML tags will be rendered as plain text (by React) instead of innerHTML.

There are columns values that are html strings, such as html links. You would need `dangerouslySetInnerHTML` and `dompurify` or use `interweave`.

kristw commented:
![image](https://user-images.githubusercontent.com/1659771/75721655-9a793400-5c8d-11ea-96ef-c9f2537510a8.png)

Could we add bottom padding for the pagination?
Also, could you share a screenshot of how does it look in Superset? I think Superset override some css and the table may look slightly different.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,238 @@
+      return val;
+    } else if (isMetric(key)) {
+      // default format '' will return human readable numbers (e.g. 50M, 33k)
+      return getNumberFormatter(format || '')(val as number);
```
Can do
```ts
import { formatNumber } from '@superset-ui/number-format';
```

```ts
return formatNumber(format || '', val as number);
```

ktmud committed:
feat(legacy-table-chart): support html in data cells

Also

1. improve height adjustment
2. add page size control to storybook
3. hide rows from later pages at initial rendering

ktmud committed:
fix(legacy-data-table): minor formatting fixes

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -35,19 +36,9 @@ if (!dt.$) {
   dt(window, $);
 }
 
+const RE_HTML_TAG = /<.*>/; // a dead simple regexp to find html tag
```
Let's rename this to `htmlTagRegex` for clarity

kristw replied:
Can make small adjustment to the regex.

`/(<([^>]+)>)/gi`

ktmud replied:
I don't think we need the `gi` flags and captured groups. Will use `/<[^>]+>/`.

nytai replied:
obligatory: https://stackoverflow.com/a/1732454

ktmud replied:
Lol. We didn't really use this to extract or remove the tags, just preemptively test whether some cell text is *potentially* HTML so we can avoid using `dangerouslySetInnerHTML`.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -80,17 +71,14 @@ export default function ReactDataTable(props: DataTableProps) {
   const maxes: { [key: string]: number } = {};
   const mins: { [key: string]: number } = {};
-  const isMetric = (key: string) => metrics.includes(key);
+  // a fater way of determine whether a key is a metric
```
typo "faster"

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -162,11 +162,10 @@ export default function ReactDataTable(props: DataTableProps) {
     } else if (typeof val === 'string') {
-      // It's fine to return raw value react will handles HTML escaping
-      return val;
+      return filterXSS(val, { stripIgnoreTag: true });
```
Could we use `dompurify` like before? 
`dompurify` seems to be more active (last updated < 1 month + 1 issue) compared to `xss` (last updated 1 year + 30 issue) with larger npm download
https://www.npmtrends.com/dompurify-vs-xss

ktmud replied:
Both have decent number of followers but `js-xss` is much faster: 
https://www.measurethat.net/Benchmarks/Show/7485/0/js-xss-vs-dompurify

There is actually a noticeable speed difference when I tested in Superset with Chrome profiling tool.

kristw replied:
I am supportive if it shows significant speed improvement. If you could note down the numbers maybe we can save that for announcement email.

BTW, is the test for `js-xss` from the link above wrong? It seems to call `sanitize`, not `filterXSS()`.
![image](https://user-images.githubusercontent.com/1659771/75822388-bc89b980-5d54-11ea-919d-550a72d9192f.png)

ktmud replied:
My bad. Use this instead: https://www.measurethat.net/Benchmarks/Show/7485/4/dompurify-vs-js-xss

js-xss is indeed slower than [sanitize.js](https://github.com/gbirke/Sanitize.js), but still faster than dompurify, though.

kristw replied:
sgtm. Thanks for being thorough.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -203,29 +202,39 @@ export default function ReactDataTable(props: DataTableProps) {
           {data.map((record, i) => (
-            <tr key={i}>
+            // hide rows after first page makes the initial render faster (less layout computation)
+            <tr key={i} style={{ display: pageLength > 0 && i >= pageLength ? 'none' : undefined }}>
```
Even with pagination, do we still create all rows in dom?

ktmud replied:
`jquery.datatables` either reads DOM attributes to collect data or we pass a `data` object with all the records. But the latter approach is actually slower because it also creates DOM nodes for all rows beforehand and calls "innerHTML" on every cell which we have no control of.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/index.ts:
```diff
@@ -16,6 +16,7 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+/// <reference path="../types/external.d.ts" />
```
Is this necessary?

ktmud replied:
This is necessary when importing from `superset-frontend` directly. Figured it's best to leave it there since it doesn't break anything anyway.

ktmud replied:
Removed to fix linting error. :D

kristw commented:
The build seems to be broken on `xss`

ktmud commented:
Addition information on `js-xss` vs `dompurify`, on the same 10k row sample data in Storybook, it takes ~12 secs for `dompurify` to finish rendering React DOM nodes and 3.5 secs for `js-xss`.

DOMPurify:
![image](https://user-images.githubusercontent.com/335541/75826804-2e660100-5d5d-11ea-92af-6c811cf6f86e.png)

xss:
![image](https://user-images.githubusercontent.com/335541/75826833-4047a400-5d5d-11ea-9a4a-6ee643050ca9.png)

ktmud committed:
chore(legacy-table): add xss dependency

Plus minor variable name and comment update

kristw commented:
* Code looks good. Some `eslint` errors breaking ci.
* Have to do something with the title.

![legacy_-___plugin_-_chart_-_table___TableChartPlugin_-_Big_Table_⋅_Storybook](https://user-images.githubusercontent.com/1659771/75830915-ee0b8080-5d66-11ea-8373-1967d6a49327.png)

ktmud committed:
fix(legacy-table): linting errors

ktmud committed:
refactor(legacy-table): more predictable metric labels

ktmud committed:
feat(legacy-table): also display title for percent metric

kristw commented on packages/superset-ui-legacy-plugin-chart-table/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@superset-ui/legacy-plugin-chart-table",
-  "version": "0.11.15",
+  "version": "0.11.5",
```
Do not modify this. This will be bumped by `lerna` when publishing. 
Also, the new version will be `0.11.16`.

ktmud replied:
Oh, I meant to revert it back to 0.11.15. Previously I changed to 0.12.0.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/package.json:
```diff
@@ -28,12 +28,14 @@
-    "datatables.net-bs": "^1.10.15",
-    "dompurify": "^2.0.6",
-    "prop-types": "^15.6.2"
+    "@superset-ui/query": "^0.12.8",
```
1) Please move this to `peerDependencies`. 

There are several `@superset-ui` packages that use singleton so we have to ensure it has only one copy, similar to how `react` works. (Although some packages do not use singleton, I made all of them always `peerDependencies` to avoid confusion whether this should be `dependencies` or `peerDependencies`)

2) `^0.12.0` might be good enough for `@superset-ui/query`?

ktmud replied:
Got it

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,252 @@
+}
+
+const { PERCENT_3_POINT } = NumberFormats;
+const isPontiallyHTML = (text: string) => /<[^>]+>/.test(text);
```
* typo `isPontiallyHTML` => `isPartiallyHTML`. 
* The regex can be defined outside of the function to avoid parsing regex every function call.

ktmud replied:
- Thanks for noticing the typo! I actually meant `isPotentiallyHTML`, which probably makes more sense than "partially"?
- I was hoping modern JS engines can already optimize for RegExp literals.

Perf test seems to suggest the uncached approach is actually faster:
https://www.measurethat.net/Benchmarks/Show/7486/1/regexp-literal-cache-or-not

I bet it has something to do with variable scope resolution.

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -0,0 +1,252 @@
+  // check whethere a key is a metric
+  const metricsLookup = new Set(aggMetrics);
+  const percentMetricsLookup = new Set(percentMetrics);
+  const isMetric = (key: string) => metricsLookup.has(key);
```
* perhaps rename `metricsLookup` to `metricsSet` and `percentMetricsLookup` to `percentMetricsSet`. 
* Could we call `metricsSet.has(key)` directly instead of using `isMetric` function which introduces additional function call?

ktmud replied:
Sure!

ktmud committed:
fix(legacy-table): typos, var names, etc

kristw commented on packages/superset-ui-legacy-plugin-chart-table/src/ReactDataTable.tsx:
```diff
@@ -116,13 +114,13 @@ export default function ReactDataTable(props: DataTableProps) {
+    if (metricsSet.has(key)) {
       // default format '' will return human readable numbers (e.g. 50M, 33k)
-      return formatNumber(format || '', val as number);
+      return formatNumber(format || SMART_NUMBER, val as number);
```
no need to explicitly set `SMART_NUMBER` here. the `NumberFormatterRegistry` has ability to set a fallback if the format string is not known by the registry. The fallback is `SMART_NUMBER` by default. 

Not hardcoding it here will make it possible to override fallback at the registry.

ktmud replied:
After installing all node_modules fresh, if I pass `''`, I will get a `this.get(...) is not a function` error from [NumberFormatterRegistry.ts](https://github.com/apache-superset/superset-ui/blob/9bb5c471e0e3d05da38929feb00ade7d0ecdc474/packages/superset-ui-number-format/src/NumberFormatterRegistry.ts#L42) . But if I use `formatNumber(format, val as number)`, I will get "type not supported" error, because [`formatNumber` does not accept `undefined`.](https://github.com/apache-superset/superset-ui/blob/9bb5c471e0e3d05da38929feb00ade7d0ecdc474/packages/superset-ui-number-format/src/NumberFormatterRegistrySingleton.ts#L12)

Should I just add `// @ts-ignore` above this line?

kristw replied:
My bad. Let me fix the bug in `@superset-ui/number-format` real quick. Give 15mins.

kristw reviewed (approved):
just a few nits

ktmud committed:
docs: update notes for metric.label

ktmud committed:
refactor(legacy-table): upgrade number-format

ktmud committed:
fix(legacy-table): upgrade dependency for storybook and yarn
